Std/Data/BitVec/Basic.lean

```diff
@@ -0,0 +1,200 @@
+/-
+Copyright (c) 2022 by the authors listed in the file AUTHORS and their
+institutional affiliations. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joe Hendrix, Wojciech Nawrocki, Leonardo de Moura
+-/
+namespace Std
+/-!
+We define bitvectors. We choose the `Fin` representation over others for its relative efficiency
+(Lean has special support for `Nat`), alignment with `UIntXY` types which are also represented
+with `Fin`, and the fact that bitwise operations on `Fin` are already defined. Some other possible
+representations are `List Bool`, `{ l : List Bool // l.length = w }`, `Fin w → Bool`.
+
+We also define many bitvector operations from the
+[`QF_BV` logic](https://smtlib.cs.uiowa.edu/logics-all.shtml#QF_BV).
+of SMT-LIBv2.
+
+TODO(Leo): match the interface to what SMT-LIB provides
+-/
+
+/--
+A bitvector of the specified width. This is represented as the underlying `Nat` number
+in both the runtime and the kernel, inheriting all the special support for `Nat`.
+-/
+structure BitVec (w : Nat) where
+  /-- Bitvector is represented as a `Fin`. -/
+  val : Fin (2^w)
+  deriving DecidableEq
+
+namespace BitVec
+
+theorem pow_two_gt_zero (w : Nat) : 2^w > 0 := by
+  apply Nat.pos_pow_of_pos; decide
+
+/-- The BitVector `i mod 2^n`. -/
+protected def ofNat (n : Nat) (i : Nat) : BitVec n :=
+  { val := Fin.ofNat' i (pow_two_gt_zero _) }
+
+/-- Given a bitvector `a`, return the underlying `Nat`. -/
+protected def toNat (a : BitVec n) : Nat :=
+  a.val.val
+
+/-- Return a bitvector `0` of size `n`.  -/
+protected def zero (n : Nat) : BitVec n :=
+  BitVec.ofNat n 0
+
+instance : Inhabited (BitVec n) where
+  default := BitVec.zero n
+
+instance : OfNat (BitVec n) i where
+  ofNat := BitVec.ofNat n i
+
+/--
+Notation for bit vector literals. `i#n` is a shorthand for `BitVec.ofNat n i`.
+TODO: is this a good notation? Do we need it? We can also write `(i : BitVec n)`
+-/
+scoped syntax:max num noWs "#" noWs num : term
+macro_rules | `($i:num#$n:num) => `(BitVec.ofNat $n $i)
+
+/-- Unexpander for bit vector literals. -/
+@[app_unexpander BitVec.ofNat] def unexpandBitVecOfNat : Lean.PrettyPrinter.Unexpander
+  | `($(_) $n:num $i:num) => `($i:num#$n:num)
+  | _                => throw ()
+
+instance : Repr (BitVec n) where
+  reprPrec a _ := repr a.val ++ "#" ++ repr n
+
+instance : ToString (BitVec n) where
+  toString a := toString (repr a)
+
+/-- Theorem for normalizing the bit vector literal representation. -/
+@[simp] theorem ofNat_eq_ofNat : @OfNat.ofNat (BitVec n) i _ = BitVec.ofNat n i := rfl
```

Comment:
Why is `ofNat` being used as the simp-normal form, instead of plain numerals?

Reply:
The proof goals are easier to understand when we make the dimensions explicit.

Reply:
Is this something that we could do with a delaborator or delaborator option? (I don't mind if we mark this as an issue to be revisited later.)

Reply:
We will keep an eye on it, and revise in the future if needed. The current solution uses only an `app_unexpander` which is super light weight.